### PR TITLE
Adds azure-infrastructure-sync with documentation and tests

### PR DESCRIPTION
**Pull request by Ba4bes**
## Description

split from https://github.com/microsoft/GitHub-Copilot-for-Azure/pull/1834
Adds the new `azure-infrastructure-sync` skill for comparing Draw.io diagrams, Bicep templates, and live Azure resources to detect drift.

## Checklist

- [ ] Tests pass locally (`cd tests && npm test`)
- [ ] **If modifying skill descriptions:** verified routing correctness with integration tests (`npm run test:skills:integration -- <skill>`)
- [ ] **If modifying skill `USE FOR` / `DO NOT USE FOR` / `PREFER OVER` clauses:** confirmed no routing regressions for competing skills
- [ ] Version bumped in skill frontmatter (if skill files changed)

### CODE REVIEW

File: plugin/skills/azure-infrastructure-sync/references/procedures/diagram-parsing.md [hunk 1]
```diff
@@ -0,0 +1,48 @@
+# Diagram Parsing Procedure
+
+> **Canonical copy:** This shared reference is duplicated across Azure infrastructure skills. Keep parallel copies synchronized when updating shared guidance.
+
+Parse a Draw.io XML file into a structured resource model. Referenced by all skills that consume Draw.io diagrams.
+
+---
+
+## Procedure
```
Comment (tmeschter) [inline plugin/skills/azure-infrastructure-sync/references/procedures/diagram-parsing.md:8]: I expect the LLM will be able to handle these instructions but they are so procedural it would be much more efficient to have a Python/Node/PowerShell/bash script that performs these steps and create the resource model, and have the LLM call that.

**Review by tmeschter — DISMISSED** [review]
I'm going to approve this, but large components of what this does could be replaced with scripts plus instructions on using the scripts.

File: tests/skills.json [hunk 2]
```diff
@@ -30,4 +31,4 @@
         "0 8 * * 2-6": "azure-deploy",
```
Comment (jongio) [inline tests/skills.json:34]: `azure-infrastructure-sync` needs to be added to the `0 12 * * 2-6` schedule. The "Skill Structure" CI check compares all skill directories against this list and fails when they don't match. Insert alphabetically between `azure-hosted-copilot-sdk` and `azure-kubernetes`.

File: tests/azure-infrastructure-sync/__snapshots__/triggers.test.ts.snap [hunk 3]
```diff
@@ -0,0 +1,109 @@
+// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing
```
Comment (jongio) [inline tests/azure-infrastructure-sync/__snapshots__/triggers.test.ts.snap:1]: Snapshots are stale - the description was updated in later commits but snapshots weren't regenerated. Run `npx jest --testPathPattern=azure-infrastructure-sync -u` and commit the updated snap file.

File: plugin/skills/azure-infrastructure-sync/SKILL.md [hunk 4]
```diff
@@ -0,0 +1,128 @@
+---
+name: azure-infrastructure-sync
+description: "Compare and synchronize infrastructure artifacts — Draw.io diagrams, Bicep templates, and live Azure resources — to detect drift and divergence. Supports diagram-to-Azure sync (quick and deep modes), Bicep-to-diagram comparison, and Bicep-to-Azure what-if analysis. WHEN: check drift, compare diagram to azure, sync bicep and diagram, bicep what-if, preview bicep changes, detect infrastructure drift, diagram drift, compare bicep deployment, does my diagram match azure, compare bicep to diagram, infrastructure drift detection. DO NOT USE FOR: generating diagrams (use azure-resource-visualizer), generating Bicep (use azure-iac-generator), pre-deployment policy checks (use azure-validate)."
```
Comment (jongio) [inline plugin/skills/azure-infrastructure-sync/SKILL.md]: The `DO NOT USE FOR` clause says "generating Bicep (use azure-iac-generator)" but `azure-iac-generator` doesn't exist in the repo. `azure-prepare` handles Bicep generation. Same phantom reference appears in `bicep-best-practices.md` line 3 ("Copied from plugin/skills/azure-iac-generator/...").

File: plugin/skills/azure-infrastructure-sync/references/bicep-whatif-workflow.md [hunk 5]
```diff
@@ -0,0 +1,43 @@
+# Bicep ↔ Azure What-If Workflow
+
+Compare Bicep templates against live Azure state without using ARM what-if. Parses Bicep into an expected resource model, queries Azure for actual state, and compares.
```
Comment (jongio) [inline plugin/skills/azure-infrastructure-sync/references/bicep-whatif-workflow.md:3]: This says "without using ARM what-if" but SKILL.md Quick Reference lists `az deployment group what-if` as the CLI fallback for Bicep-to-Azure mode. These contradict each other - pick one approach and align both files.

File: plugin/skills/azure-infrastructure-sync/references/procedures/bicep-parsing.md [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+## Procedure
+
+### 1. Read Parameter Values
+
+Read the `.bicepparam` file and extract all parameter values. These are needed to resolve name expressions in resource blocks.
+
+### 2. Read Template Files
+
+Read `main.bicep` and parse all `module` declarations to find referenced Bicep files. Module paths are relative to the file containing the `module` statement — they may reference files outside `modules/` (e.g., `'../shared/networking.bicep'` or `'br:myregistry.azurecr.io/bicep/networking:v1'`). Read each resolved module file recursively to discover nested module references.
+
```
Comment (jongio) [inline plugin/skills/azure-infrastructure-sync/references/procedures/bicep-parsing.md:19]: Step 2 treats `br:` registry modules like local files ("read each resolved module file recursively"), but `br:` references aren't file paths - they need `bicep restore` first. Add a note that registry modules need restoration or should be flagged as "external - can't resolve locally."

File: plugin/skills/azure-infrastructure-sync/references/azure-stencil-mapping.json [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+    },
+
+    "_comment_wrongCategory": "These icons are in unexpected categories — the LLM would guess wrong",
+    "Microsoft.Web/serverFarms": {
+      "imagePath": "img/lib/azure2/app_services/App_Service_Plans.svg",
+      "label": "App Service Plan",
+      "defaultWidth": 50,
+      "defaultHeight": 50,
+      "note": "CATEGORY: app_services — distinct icon from App_Services.svg. Do NOT use compute/ (file does not exist there)."
+    },
```
Comment (jongio) [inline plugin/skills/azure-infrastructure-sync/references/azure-stencil-mapping.json:63]: `Microsoft.Web/serverFarms` uses capital F, but ARM returns `Microsoft.Web/serverfarms` (lowercase). Same casing mismatch in `azure-deployment-verification.md` with `ApplicationGatewayWebApplicationFirewallPolicies`. If lookups are case-sensitive, this causes matching failures. Either standardize on ARM convention or document that lookups must be case-insensitive.

File: tests/azure-infrastructure-sync/triggers.test.ts [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+      'triggers on: "%s"',
+      (prompt) => {
+        const result = triggerMatcher.shouldTrigger(prompt);
+        expect(result.triggered).toBe(true);
+      }
+    );
+  });
+
+  describe("Should NOT Trigger", () => {
+    const shouldNotTriggerPrompts: string[] = [
```
Comment (jongio) [inline tests/azure-infrastructure-sync/triggers.test.ts:45]: The "Should NOT Trigger" tests only use generic unrelated prompts (weather, poems). Add negative cases for competitor skills: "generate Bicep for my architecture", "validate my Bicep deployment", "create a diagram of my Azure resources". These test that the DO NOT USE FOR boundaries actually work.

**Review by jongio — DISMISSED** [review]
Two CI checks are failing - the scheduled integration test list in `skills.json` is missing this skill, and the trigger snapshots are stale. Beyond CI, there are a few doc consistency issues worth fixing before merge.

The skill architecture is solid - clean progressive disclosure with SKILL.md routing to 4 workflow references, consistent drift report format, and thoughtful normalization rules for deep mode.

File: plugin/skills/azure-infrastructure-sync/SKILL.md (hunk 4 above)
Comment (Ba4bes) [reply plugin/skills/azure-infrastructure-sync/SKILL.md]: This made sense for the previous bulk-PR, but not for this structure. So it is removed.

File: plugin/skills/azure-infrastructure-sync/references/azure-deployment-verification.md [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+
+---
+
+## 1. SKU Dependency Rules
+
+Certain SKUs require companion resources or specific configurations. Missing these causes deployment failures.
+
+### 1.1 Application Gateway WAF_v2 requires WAF Policy
+- **Applies to**: `Microsoft.Network/applicationGateways` with `sku.tier == 'WAF_v2'`
+- **Rule**: A `Microsoft.Network/ApplicationGatewayWebApplicationFirewallPolicies` resource MUST exist and be referenced via the `firewallPolicy.id` property on the Application Gateway.
```
[the remark this answers is not in the document]
Comment (Ba4bes) [reply plugin/skills/azure-infrastructure-sync/references/azure-deployment-verification.md]: Fixed capitalisation as suggested

File: plugin/skills/azure-infrastructure-sync/references/azure-stencil-mapping.json (hunk 7 above)
[the remark this answers is not in the document]
Comment (Ba4bes) [reply plugin/skills/azure-infrastructure-sync/references/azure-stencil-mapping.json:63]: Fixed capitalisation as suggested

Comment (Ba4bes) [reply plugin/skills/azure-infrastructure-sync/references/azure-stencil-mapping.json:63]: Fixed capitalisation as suggested

File: plugin/skills/azure-infrastructure-sync/references/bicep-whatif-workflow.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (Ba4bes) [reply plugin/skills/azure-infrastructure-sync/references/bicep-whatif-workflow.md:3]: Fixed description, it should indeed not use what-if

Comment (Ba4bes) [reply plugin/skills/azure-infrastructure-sync/references/bicep-whatif-workflow.md:3]: Fixed description, it should indeed not use what-if

File: plugin/skills/azure-infrastructure-sync/references/procedures/bicep-parsing.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (Ba4bes) [reply plugin/skills/azure-infrastructure-sync/references/procedures/bicep-parsing.md]: Fixed description as suggested

Comment (Ba4bes) [reply plugin/skills/azure-infrastructure-sync/references/procedures/bicep-parsing.md:19]: Fixed description as suggested

File: plugin/skills/azure-infrastructure-sync/references/procedures/diagram-parsing.md (hunk 1 above)
Comment (Ba4bes) [reply plugin/skills/azure-infrastructure-sync/references/procedures/diagram-parsing.md:8]: I agree that would be a great option for all procedures that don't require the Azure connections. But as this PR is pretty big as it is, I suggest to leave this as a possible todo for a later time. The procedures do work as expected with the description

File: plugin/skills/azure-infrastructure-sync/references/procedures/resource-matching.md [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+
+Apply rules in order. Stop at the first match.
+
+| # | Rule | Result |
+|---|------|--------|
+| 1 | Type AND name match (case-insensitive) | **In Sync** |
+| 2 | Type matches AND only ONE resource of that type exists in EACH model | **In Sync (name differs)** — report both names |
+| 3 | Type matches AND multiple exist → name has case-insensitive substring match or Levenshtein distance ≤ 3 | **In Sync (name differs)** |
+| 4 | Exists in Model A only | **Model A Only** (e.g., "Diagram Only", "Bicep Only", "Azure Only") |
+| 5 | Exists in Model B only | **Model B Only** |
```
[the remark this answers is not in the document]
Comment (Ba4bes) [reply plugin/skills/azure-infrastructure-sync/references/procedures/resource-matching.md:17]: I think this is a bug in Copilot as I don't see the double leading pipe.

File: tests/azure-infrastructure-sync/__snapshots__/triggers.test.ts.snap (hunk 3 above)
Comment (Ba4bes) [reply tests/azure-infrastructure-sync/__snapshots__/triggers.test.ts.snap:1]: I tried to run that command, but it kept failing. I asked Copilot to look at it and it ended up changing all the tests themselves, which doesn't seem like the right scope for this change. I have now let Copilot change the file manually

File: tests/azure-infrastructure-sync/triggers.test.ts (hunk 8 above)
Comment (Ba4bes) [reply tests/azure-infrastructure-sync/triggers.test.ts:45]: I tried to use negative cases like that, but they were not passing the tests because they were too similar to the actual skill. I have now tried to copy the logic that I saw in the other Skills. Would this work, or should it still be more specific?

File: tests/skills.json (hunk 2 above)
Comment (Ba4bes) [reply tests/skills.json:34]: added to the schedule

File: tests/skills.json [hunk 11]
```diff
@@ -30,4 +31,4 @@
...
         "0 12 * * 2-6": "appinsights-instrumentation,azure-ai,azure-aigateway,azure-cloud-migrate,azure-compliance,azure-compute,azure-cost,azure-diagnostics,azure-enterprise-infra-planner,azure-hosted-copilot-sdk,azure-kubernetes,azure-kusto,azure-messaging,azure-prepare,azure-quotas,azure-rbac,azure-resource-lookup,azure-resource-visualizer,azure-storage,azure-upgrade,azure-validate,entra-app-registration"
```
[the remark this answers is not in the document]
Comment (Ba4bes) [reply tests/skills.json]: added to the schedule

File: plugin/skills/azure-infrastructure-sync/references/azure-resource-configs.md [hunk 12]
```diff
@@ -0,0 +1,198 @@
...
+
+Each resource type section includes:
+- **MCP Tool**: Primary Azure MCP tool to query (if available)
+- **Fallback**: CLI command for types without a dedicated MCP tool
+- **Field Paths**: JSON paths in the ARM response
+
+### Microsoft.Web/serverfarms
+
+**MCP Tool**: `mcp_azure_appservice`
+**Fallback**: `az appservice plan show --ids <resourceId> -o json`
```
[the remark this answers is not in the document]
Comment (Ba4bes) [reply plugin/skills/azure-infrastructure-sync/references/azure-resource-configs.md:41]: Changed into mcp_azure_mcp_ as suggested

File: plugin/skills/azure-infrastructure-sync/references/procedures/azure-authentication.md [hunk 13]
```diff
@@ -0,0 +1,34 @@
...
+
+> **Canonical copy:** This shared reference is duplicated across Azure infrastructure skills. Keep parallel copies synchronized when updating shared guidance.
+
+Canonical procedure for verifying Azure session before any Azure operations. Referenced by all skills that interact with Azure.
+
+---
+
+## Procedure
+
+1. **Verify session**: Run `az account show` or `Get-AzContext` in the terminal, OR attempt a lightweight Azure MCP call (e.g., `mcp_azure_subscription_list`)
```
[the remark this answers is not in the document]
Comment (Ba4bes) [reply plugin/skills/azure-infrastructure-sync/references/procedures/azure-authentication.md]: Changed into mcp_azure_mcp_ as suggested

File: plugin/skills/azure-infrastructure-sync/references/azure-resource-configs-compute.md [hunk 14]
```diff
@@ -0,0 +1,108 @@
+# Azure Compute & Container Resource Configuration Reference
+
+ARM property retrieval mapping for compute and container resources. Used by [azure-resource-configs.md](azure-resource-configs.md).
+
+---
+
+### Microsoft.Compute/virtualMachines
+
+**MCP Tool**: `mcp_azure_compute`
+**Fallback**: `az vm show --ids <resourceId> -o json`
```
[the remark this answers is not in the document]
Comment (Ba4bes) [reply plugin/skills/azure-infrastructure-sync/references/azure-resource-configs-compute.md]: Changed into mcp_azure_mcp_ as suggested

File: plugin/skills/azure-infrastructure-sync/SKILL.md [hunk 15]
```diff
@@ -0,0 +1,128 @@
+---
+name: azure-infrastructure-sync
+description: "Compare and synchronize infrastructure artifacts — Draw.io diagrams, Bicep templates, and live Azure resources — to detect drift and divergence. Supports diagram-to-Azure sync (quick and deep modes), Bicep-to-diagram comparison, and Bicep-to-Azure what-if analysis. WHEN: check drift, compare diagram to azure, sync bicep and diagram, bicep what-if, preview bicep changes, detect infrastructure drift, diagram drift, compare bicep deployment, does my diagram match azure, compare bicep to diagram, infrastructure drift detection. DO NOT USE FOR: generating diagrams (use azure-resource-visualizer), pre-deployment policy checks (use azure-validate)."
+license: MIT
+metadata:
+  author: Microsoft
+  version: "1.0.0"
```
[the remark this answers is not in the document]
Comment (Ba4bes) [reply plugin/skills/azure-infrastructure-sync/SKILL.md]: Changed as suggested

File: plugin/skills/azure-infrastructure-sync/SKILL.md [hunk 16]
```diff
@@ -0,0 +1,128 @@
...
+| Bicep formatting | Optional | `filePath` | Normalize Bicep before parsing or review |
+| Draw.io diagram access/update | Optional | diagram path, XML payload | Read or update diagram artifacts during sync |
+
+| CLI fallback | When to use | Common parameters | Notes |
+|---|---|---|---|
+| `az resource list` | MCP resource listing unavailable | `--resource-group`, `--subscription` | Baseline inventory for quick mode |
+| `az resource show` | Need raw resource properties | `--ids` or `--name`, `--resource-group`, `--resource-type` | Useful for deep comparisons |
+| `az graph query` | Cross-resource-group or broad discovery | `-q`, `--subscriptions` | Good for large scopes |
+| `az deployment group what-if` | Validate Bicep against a resource group | `--resource-group`, `--template-file`, `--parameters` | Preferred CLI preview for RG deployments |
+| `az deployment sub what-if` | Validate subscription-scope Bicep | `--location`, `--template-file`, `--parameters` | Use for subscription deployments |
```
[the remark this answers is not in the document]
Comment (Ba4bes) [reply plugin/skills/azure-infrastructure-sync/SKILL.md]: Removed and replaced with bicep build

File: plugin/skills/azure-infrastructure-sync/references/azure-resource-model.md [hunk 17]
```diff
@@ -0,0 +1,104 @@
...
+|------|-------------|---------|
+| `contains` | Parent contains child resource | VNet contains Subnet |
+| `connects` | Network or data flow connection | VM connects to Storage Account |
+| `depends` | Deployment dependency | App Service depends on App Service Plan |
+| `peers` | Bidirectional peering | VNet peers with VNet |
+| `secures` | Security association | NSG secures Subnet |
+| `routes` | Traffic routing | Load Balancer routes to VM |
+
+## Usage by Skill
+
```
Comment (jongio) [inline plugin/skills/azure-infrastructure-sync/references/azure-resource-model.md]: **[MEDIUM]** `Usage by Skill` references 5 skills that don't exist in the repo: `azure-to-bicep`, `azure-to-diagram`, `diagram-to-bicep`, `sketch-to-diagram`, `bicep-policy-check`. Same class of issue as the `azure-iac-generator` ref we already fixed in SKILL.md.

The intro at L6 also references "AzVerify skills" as though it's a real product family - it isn't.

Rewrite this section to describe the workflow modes of *this* skill (`azure-infrastructure-sync`): diagram-azure-sync, bicep-diagram-sync, and bicep-whatif. Drop the skills that aren't in the repo.

File: plugin/skills/azure-infrastructure-sync/references/procedures/diagram-parsing.md [hunk 18]
```diff
@@ -0,0 +1,48 @@
+# Diagram Parsing Procedure
+
+> **Canonical copy:** This shared reference is duplicated across Azure infrastructure skills. Keep parallel copies synchronized when updating shared guidance.
```
Comment (jongio) [inline plugin/skills/azure-infrastructure-sync/references/procedures/diagram-parsing.md]: **[LOW]** "This shared reference is duplicated across Azure infrastructure skills" - there are no other infrastructure skills to keep in sync with. Same note appears in 4 other reference files. These are leftover from the #1834 bulk PR and misleading here. Either remove the canonical copy notes or rephrase to say this is a standalone reference used by `azure-infrastructure-sync`.

**Review by jongio — COMMENTED** [review]
Previous comments addressed - nice work on the fixes. Two docs issues from the bulk PR split:

- references/azure-resource-model.md:58 - `Usage by Skill` lists skills that don't exist in the repo
- 5 reference files have `Canonical copy` headers referencing sibling skills that aren't part of this PR

Neither is blocking, but both would confuse anyone reading these docs for the first time.

File: plugin/skills/azure-infrastructure-sync/references/azure-resource-model.md (hunk 17 above)
Comment (Ba4bes) [reply plugin/skills/azure-infrastructure-sync/references/azure-resource-model.md]: Removed references in this file and other files

File: plugin/skills/azure-infrastructure-sync/references/procedures/diagram-parsing.md (hunk 18 above)
Comment (Ba4bes) [reply plugin/skills/azure-infrastructure-sync/references/procedures/diagram-parsing.md]: Removed references in this file and other files

File: plugin/skills/azure-infrastructure-sync/SKILL.md [hunk 19]
```diff
@@ -0,0 +1,125 @@
...
+All modes produce a consistent drift report:
+
+```
+## Infrastructure Drift Report — <Mode>
+
+| # | Resource | Type | Status |
+|---|----------|------|--------|
+| 1 | vnet-web | Microsoft.Network/virtualNetworks | ✅ In Sync |
+| 2 | app-api | Microsoft.Web/sites | ⬜ Diagram Only |
+| 3 | redis-cache | Microsoft.Cache/redis | 🔷 Azure Only |
```
[the remark this answers is not in the document]
Comment (Ba4bes) [reply plugin/skills/azure-infrastructure-sync/SKILL.md]: updated to more generic example

File: plugin/skills/azure-infrastructure-sync/references/procedures/azure-authentication.md [hunk 20]
```diff
@@ -0,0 +1,31 @@
+# Azure Authentication Check
+
+
+---
+
+## Procedure
+
+1. **Verify session**: Run a lightweight Azure MCP call (e.g., `mcp_azure_mcp_subscription_list`), or run `az account show` or `Get-AzContext` in the terminal
```
[the remark this answers is not in the document]
Comment (Ba4bes) [reply plugin/skills/azure-infrastructure-sync/references/procedures/azure-authentication.md:8]: this has already been fixed

File: tests/azure-infrastructure-sync/triggers.test.ts [hunk 21]
```diff
@@ -0,0 +1,103 @@
...
+      "Write a TypeScript function to call a REST API",
+      "Implement retry logic for my HTTP client",
+
+      // Azure cost management (use azure-cost)
+      "How much am I spending on Azure this month?",
+      "Show me my Azure billing summary",
+
+      // Azure service questions that are out of scope for infra drift/sync
+      "Troubleshoot Azure Redis latency",
+      "Azure Redis pricing",
```
[the remark this answers is not in the document]
Comment (Ba4bes) [reply tests/azure-infrastructure-sync/triggers.test.ts]: This looks like it is becoming a loop, as these lines were added on request of Copilot. Trying the suggested update one more time

File: plugin/skills/azure-infrastructure-sync/references/procedures/azure-authentication.md [hunk 22]
```diff
@@ -0,0 +1,31 @@
...
+   You need an active Azure session.
+
+   **Option A — Azure CLI:**
+   az login
+
+   **Option B — Azure PowerShell:**
+   Connect-AzAccount
+
+   After authenticating, run this skill again.
+   ```
```
[the remark this answers is not in the document]
Comment (Ba4bes) [reply plugin/skills/azure-infrastructure-sync/references/procedures/azure-authentication.md]: updated with "markdown"

File: plugin/skills/azure-infrastructure-sync/SKILL.md [hunk 23]
```diff
@@ -0,0 +1,125 @@
...
+
+```
+## Infrastructure Drift Report — <Mode>
+
+| # | Resource | Type | Status |
+|---|----------|------|--------|
+| 1 | vnet-web | Microsoft.Network/virtualNetworks | ✅ In Sync |
+| 2 | app-api | Microsoft.Web/sites | ⬜ Diagram Only |
+| 3 | vm01 | Microsoft.Compute/virtualMachines | 🔷 Azure Only |
+
```
[the remark this answers is not in the document]
Comment (Ba4bes) [reply plugin/skills/azure-infrastructure-sync/SKILL.md:112]: Removed another reference to redis, rebuild the tests. (other references are relevant)

File: plugin/skills/azure-infrastructure-sync/SKILL.md [hunk 24]
```diff
@@ -0,0 +1,125 @@
...
+    └─► [bicep-whatif-workflow.md]
+```
+
+## Workflow References
+
+- **Diagram ↔ Azure (quick)**: [diagram-azure-sync-workflow.md](references/diagram-azure-sync-workflow.md) — Resource-level existence comparison. Reports resources present in diagram only, Azure only, or both.
+- **Diagram ↔ Azure (deep)**: [diagram-azure-sync-deep-workflow.md](references/diagram-azure-sync-deep-workflow.md) — Property-level comparison with normalization rules. Reports configuration drift with severity.
+- **Bicep ↔ Diagram**: [bicep-diagram-sync-workflow.md](references/bicep-diagram-sync-workflow.md) — Compares Bicep resource definitions against diagram resource model. Offers selective resolution.
+- **Bicep ↔ Azure (drift comparison)**: [bicep-whatif-workflow.md](references/bicep-whatif-workflow.md) — Compares Bicep-declared state against live Azure resource state for drift analysis. This is not a deployment readiness or policy validation workflow. Reports Create/Modify/Delete/No Change.
+
```
[the remark this answers is not in the document]
Comment (Ba4bes) [reply plugin/skills/azure-infrastructure-sync/SKILL.md:95]: added a numbered list

File: tests/azure-infrastructure-sync/__snapshots__/triggers.test.ts.snap (hunk 3 above)
Comment (Ba4bes) [reply tests/azure-infrastructure-sync/__snapshots__/triggers.test.ts.snap:1]: It worked now, so snapshots have been rebuild

File: plugin/skills/azure-infrastructure-sync/references/procedures/resource-matching.md [hunk 25]
```diff
@@ -0,0 +1,41 @@
...
+
+## Matching Algorithm
+
+Apply rules in order. Stop at the first match.
+
+| # | Rule | Result |
+|---|------|--------|
+| 1 | Type AND name match (case-insensitive) | **In Sync** |
+| 2 | Type matches AND only ONE resource of that type exists in EACH model | **In Sync (name differs)** — report both names |
+| 3 | Type matches AND multiple exist → name has case-insensitive substring match or Levenshtein distance ≤ 3 | **In Sync (name differs)** |
```
Comment (jongio) [inline plugin/skills/azure-infrastructure-sync/references/procedures/resource-matching.md]: Rule 3's Levenshtein distance threshold of 3 is aggressive for short resource names. Names like `vm-1` and `db-1` have distance 2, so they'd match as "In Sync (name differs)" despite being completely different resources. Consider gating this rule on minimum name length (e.g., only apply fuzzy matching when both names are >= 8 characters), or lowering the threshold to 1-2.

**Review by jongio — COMMENTED** [review]
Clean skill structure with good progressive disclosure across reference files. The routing boundaries (DO NOT USE FOR azure-resource-visualizer, azure-validate) are well-defined. One concern on the fuzzy matching algorithm.

File: tests/azure-infrastructure-sync/triggers.test.ts (hunk 8 above)
Comment (jongio) [reply tests/azure-infrastructure-sync/triggers.test.ts:45]: The existing negative tests follow the same generic-prompts pattern used by other skills (azure-deploy, etc.), so this isn't a regression from repo conventions. That said, the point about boundary tests is valid - the TriggerMatcher's keyword-overlap behavior means prompts like "generate Bicep for my architecture" or "create a diagram of my Azure resources" could hit enough shared keywords to trigger. If those prompts currently trigger (as @Ba4bes found), that's worth documenting as known limitations in the test file with `test.skip` or a comment, rather than silently omitting them. It surfaces the routing boundary issue for future TriggerMatcher improvements.

File: plugin/skills/azure-infrastructure-sync/references/procedures/resource-matching.md (hunk 25 above)
Comment (Ba4bes) [reply plugin/skills/azure-infrastructure-sync/references/procedures/resource-matching.md]: Updated as suggested

**Review by jongio — COMMENTED** [review]
Addresses my Levenshtein threshold concern on resource-matching.md. The distance <= 2 with >= 8 character minimum prevents false matches on short names like `vm1` vs `vm2`. The Workflow section addition is clear and follows good progressive disclosure.

Cleanup of 'canonical copy' notes, casing fix on `Microsoft.Insights`, and removal of the phantom `azure-iac-generator` reference all look good.

The 2 new negative trigger tests are welcome. The boundary-case tests for competitor-skill routing (prompts that share keywords with azure-validate, azure-prepare, etc.) remain open in the triggers thread, though that's more of a TriggerMatcher design question than a bug here.

**Review by tmeschter — DISMISSED** [review]

**Comment by Ba4bes** [comment]
- the first test is failing because the overall token amount for all skills are now above a maximum amount. This seems to be something bigger than just this PR, since it is about all skills. @tmeschter  can you give some guidance on if this is something that needs action from me?
- the other test failed because a word was missing from the snapshot. Fixed it and tests are now successful locally

File: plugin/skills/azure-infrastructure-sync/references/bicep-diagram-sync-workflow.md [hunk 26]
```diff
@@ -0,0 +1,19 @@
+# Bicep ↔ Diagram Sync Workflow
+
+> **Canonical copy:** This shared reference is duplicated across Azure infrastructure skills. Keep parallel copies synchronized when updating shared guidance.
```
Comment (jongio) [inline plugin/skills/azure-infrastructure-sync/references/bicep-diagram-sync-workflow.md:3]: This `Canonical copy` header was removed from other reference files per my earlier review but still remains here. Since there aren't parallel copies across sibling skills, consider removing it.

**Review by jongio — COMMENTED** [review]
Re-reviewed after rebase (branch now includes 118 commits from main since my April 22 review).

Prior feedback was addressed: "Usage by Skill" section removed from resource-model.md, `br:` registry modules now properly flagged as external, Levenshtein threshold tightened (distance <= 2, min 8 chars), `serverfarms` casing fixed in stencil mapping, skill added to integration test schedule.

Two items still open:

1. Two reference files and the stencil mapping JSON still have "Canonical copy" headers referencing "sibling infrastructure skills" that don't exist in this repo. See inline comments.

2. The branch has merge conflicts that need resolving before this can merge.

File: plugin/skills/azure-infrastructure-sync/references/azure-deployment-verification.md [hunk 27]
```diff
@@ -0,0 +1,151 @@
+# Azure Deployment Verification Rules
+
+> **Canonical copy:** This shared reference is duplicated across Azure infrastructure skills. Keep parallel copies synchronized when updating shared guidance.
```
Comment (jongio) [inline plugin/skills/azure-infrastructure-sync/references/azure-deployment-verification.md:3]: Same orphaned `Canonical copy` header. Should be removed to match the cleanup done in other reference files.

**Review by jongio — DISMISSED** [review]
Solid change. The esbuild bundling approach correctly solves the Azure Artifacts feed dependency problem, and the generated TypeScript modules eliminate runtime `fs` reads for Bun standalone executables. All 14 CI checks pass (including integration tests on both Ubuntu and Windows), confirming the bundle works as intended.

A note on the Copilot bot's inline comments: the claim that `packages: 'bundle'` is needed is incorrect. With `bundle: true` and `platform: 'node'`, esbuild bundles npm packages by default; only Node built-ins (`fs`, `path`, etc.) are auto-externalized. CI passing validates this. The `collectRunIde` function the bot references also doesn't exist in this file.

One minor observation: `loadTypeRegistry` now always returns `CURSOR_PROTO_TYPES` (never null), but its return type is still `TypeRegistry | null` and callers still branch on `null`. Not blocking since the code is functionally correct, just makes the null path unreachable.

**Review by jongio — COMMENTED** [review]
Incremental review after the latest round of fixes. Most feedback from prior reviews has been addressed:

- skills.json updated (skill list + scheduled run)
- Snapshots regenerated with the latest keywords (including `vnet`)
- Non-existent skill references removed from azure-resource-model.md
- MCP tool naming normalized to `mcp_azure_mcp_*` convention
- Resource matching Levenshtein threshold refined (distance <= 2, names >= 8 chars)
- Bicep-whatif workflow description clarified

Two documentation nits remain (flagged via inline comments): `bicep-diagram-sync-workflow.md` and `azure-deployment-verification.md` still carry orphaned 'Canonical copy' headers referencing sibling infrastructure skills that don't exist. Removing those two lines would complete the cleanup done in the other reference files.

Not blocking on those. The skill structure, routing, tests, and workflow docs are all solid.
